src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoEnumTrailingCommaCheck.java

```diff
@@ -0,0 +1,128 @@
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2019 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+/**
+ * <p>
+ *  Checks that enum definition does not contain a trailing comma.
+ * </p>
+ * <pre>
+ * enum Foo1 {
+ *   FOO,
+ *   BAR;
+ * }
+ * </pre>
+ * <p>
+ *  The check demands that there should not be any comma after last constant in
+ *  enum definition.
+ * </p>
+ * <pre>
+ * enum Foo1 {
+ *   FOO,
+ *   BAR, //violation
+ * }
+ * </pre>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name=&quot;NoEnumTrailingComma&quot;/&gt;
+ * </pre>
+ * <p>
+ * Which results in the following violations:
+ * </p>
+ * <pre>
+ * enum Foo1 {
+ *   FOO,
+ *   BAR; //OK
+ * }
+ * enum Foo2 {
+ *   FOO,
+ *   BAR //OK
+ * }
+ * enum Foo3 {
+ *   FOO,
+ *   BAR, //violation
+ * }
+ * enum Foo4 {
+ *   FOO, // OK
+ *   BAR, // violation
+ *   ;
+ * }
+ * enum Foo5 {
+ *   FOO, // OK
+ *   BAR,; // violation
+ * }
+ * enum Foo6 { FOO, BAR,; } // violation
+ * enum Foo7 { FOO, BAR, } // violation
+ * enum Foo8 {
+ *   FOO, // OK
+ *   BAR // OK
+ *   ;
+ * }
+ * enum Foo9 { FOO, BAR; } // OK
+ * enum Foo10 { FOO, BAR } // OK
+ * </pre>
+ *
+ * @since 8.28
+ */
+@StatelessCheck
+public class NoEnumTrailingCommaCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "no.enum.trailing.comma";
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {TokenTypes.ENUM_DEF};
+    }
+
+    @Override
+    public void visitToken(DetailAST detailAST) {
```

Comment:
This can be simplified to
```
    public void visitToken(DetailAST detailAST) {
        final DetailAST objBlock = detailAST.findFirstToken(TokenTypes.OBJBLOCK);
        TokenUtil.forEachChild(objBlock, TokenTypes.COMMA, comma -> {
            final int type = comma.getNextSibling().getType();
            if (type != TokenTypes.VARIABLE_DEF && type != TokenTypes.ENUM_CONSTANT_DEF) {
                log(comma, MSG_KEY);
            }
        });
    }
```
All we need to do is check that each comma is followed by a enum field or regular field (see my comment for complex enums).

Reply:
Done